packages/vue-components/src/Navbar.vue

```diff
@@ -77,10 +77,41 @@
       toggleCollapse (e) {
         e && e.preventDefault()
         this.collapsed = !this.collapsed
+      },
+      highlightCurrentPage() {
+        function hasMatchingUrl(elements, currPage) {
+          if (!elements || !elements.children) {
+            return false;
+          }
+          // Only check <a> leaf nodes
+          if (elements.children.length === 0) {
+            if (elements.href) {
+              if (elements.href === currPage) {
+                return true;
+              }
+            }
+            return false;
+          }
+          // otherwise, check all children recursively
+          return Array.from(elements.children).some(node => hasMatchingUrl(node, currPage));
+        }
+        document.addEventListener('DOMContentLoaded', () => {
+          const navLinks = document.querySelectorAll('.navbar .navbar-nav .nav-link');
+          if (!navLinks) {
+            return;
+          }
+          const currPage = window.location.href;
```

Comment:
the detection does not work currently for implicit `index.html` links (e.g. `markbind.org/userGuide` / `markbind.org/userGuide/`-> points to `markbind.org/userGuide/index.html`)

you'll need to normalize both the `href` and `window.location.href`s when checking

Reply:
should we make the checking "generalizable" as well?

e.g. if the link is `markbind.org/userGuide/index.html`,

`userGuide/index.html`
`userGuide/formattingContents.html`
`userGuide/formattingContents.html`

would all match

Since the links in the navbar are usually only a small subset

---

On the other hand, this may make the checking ambiguous though; we'll need to define the cases clearly and see if there's any such case before proceeding

What do you think? @KendrickAng

Reply:
> 
> 
> the detection does not work currently for implicit `index.html` links (e.g. `markbind.org/userGuide` / `markbind.org/userGuide/`-> points to `markbind.org/userGuide/index.html`)
> 
> you'll need to normalize both the `href` and `window.location.href`s when checking

What do you mean by "implicit index.html links"? I thought that all links in the navbar are well-defined, i.e we hardcode them in the `header.md` in layouts.

Didn't think of the normalisation part, I'll do that!

> 
> 
> should we make the checking "generalizable" as well?
> 
> e.g. if the link is `markbind.org/userGuide/index.html`,
> 
> `userGuide/index.html`
> `userGuide/formattingContents.html`
> `userGuide/anotherfolder/formattingContents.html`
> 
> would all match
> 
> Since the links in the navbar are usually only a small subset
> 
> On the other hand, this may make the checking ambiguous though; we'll need to define the cases clearly and see if there's any such case before proceeding
> 
> What do you think? @KendrickAng

In that case, the navbar would need to maintain some sort of state (luckily we use vue components, which has state)
But then this brings up another issue - what if the side nav href doesn't match the top nav href? For example link is `markbind.org/userGuide/index.html`, side nav has `userGuide/formattingContents.html` and ` anotherName/anotherfolder/formattingContents.html`.

In that case, there wouldn't be a way of knowing that those links have the same parent `markbind.org/userGuide/index.html`. Do you have an idea how to overcome this?

Reply:
> What do you mean by "implicit index.html links"?

>`markbind.org/userGuide` / `markbind.org/userGuide/`-> points to `markbind.org/userGuide/index.html`

links like these

e.g. the main page of netlify preview only highlights `home` if the address is https://deploy-preview-1379--markbind-master.netlify.app/index.html (vs  https://deploy-preview-1379--markbind-master.netlify.app/ which should also highlight it)

> In that case, the navbar would need to maintain some sort of state

state wouldn't be needed since we're checking once-per-page only, just some extra logic around how you match the current address to the `href`s 

> what if the side nav href doesn't match the top nav href?

hmm "generalizable" was a bad choice of words (our site nav already has highlighting)

I meant to "relax" the search conditions - so in the above example, those 3 addresses are _browser addresses_, the one before that is the navbar link

For now I'm thinking simply:

If after link normalization (changing links like `userGuide/` into `userGuide/index.html`), the link points to some `index.html`, we could check if `window.location.href` is a _sibling_ or _child_ of this. (e.g. `userGuide/formattingContents.html` or `userGuide/anotherfolder/somefile.html`) If so we highlight the link

If after link normalization, it points to something other than some index.html, we do the exact equality check already present in this pr

Reply:
oh so normalization meant adding an `index.html` to links that don't end with `.html`, I thought you mean something like this: thttps://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize

Reply:
yup

Reply:
Actually, I realise that this checking of sibling or child may not work well, given that there may be cases like this, where multiple nav links have a common prefix:
![image](https://user-images.githubusercontent.com/38814428/97796331-9e5dbb80-1c4b-11eb-88b8-5a3bb00835b5.png)
I think we could employ a stricter check (the dfs check) for an exact matching, and if that doesn't work, use the relaxed checking you suggested here. I believe that works for the current markbind docs and the example I listed here

Reply:
Hmm we could do some sort of priority matching here -> the more "groups" (`/`) the link has, the more priority we give to it. And then we only highlight the one with the highest priority

I think this could actually be desired behaviour on the other hand though
e.g. if you have this dropdown structure
```
userGuide/
  topic1.html
```
both the main dropdown link and `topic1.html` is highlighted

Any thoughts 😮

Reply:
Another alternative is to allow some additional attributes to control the highlighting behaviour:

e.g.
```
<a href="..."
  highlight-on="none / exact / child / sibling / child or sibling" // default this to "child or sibling" unless specified
  highlight-link="..." // we can even highlight non-links this way (another pr)
  highlight-class="..." // maybe even allow the user to specify this, but default it to something reasonable (another pr maybe)
  />
```

Reply:
> when we match on groups (/), clicking on topic3a.html causes topic1.html's link to be incorrectly highlighted instead, since the relaxed checking checks for a common prefix (contents in this case)

> If after link normalization, it points to something other than some index.html, we do the exact equality check already present in this pr

Do you mean if the address bar is `contents/topic3a.html`, `contents/topic1.html` would also be highlighted?
My suggestion here was to do _only_ exact equality checks for such cases (non `index.html`s). (versus falling back to the prefix match at all)

But this behaviour may be rather obscure (falling back in the case of `index.html`s, not falling back at all otherwise).

https://github.com/MarkBind/markbind/pull/1379#discussion_r515591206
Hence we could always fall back, but provide an alternative  to control the mode of checking

> However, I believe that it would be quite rare where a user does something like that though (not using the exact url, like in markbind docs)

Hmm not sure I understood this, what does "not using the exact url" refer to? (wouldn't links as specified by `<a href="..." />` always point to an exact url?)

Reply:
I think we may be talking about different issues here... The problem was that since we could have something like `topic1.html` and `topic3a.html` having the same prefix `contents`, when we match on groups (/), clicking on `topic3a.html` causes `topic1.html`'s link to be highlighted instead, since the relaxed checking checks for a common prefix (contents in this case)

Hence I thought of doing an exact URL match first, then if that doesn't work, do a match on the highest-priority URL (with the most groups matched).

Reply:
>Hmm not sure I understood this, what does "not using the exact url" refer to? (wouldn't links as specified by <a href="..." /> always point to an exact url?)

I was talking about the case where the href of an anchor tag doesn't match the *end result* of `window.location.href` once it is clicked, for example:

1. Suppose we have this entry in `layouts/devGuide/header.md` <div tags="environment--dg"><li><a href="{{baseUrl}}/devGuide/index.html" class="nav-link">DEVELOPER GUIDE</a></li></div>`
1. However, `devGuide/index.md` contains this code:
><script>
> window.location.href = "devGuide.html"
></script>
1. So when the anchor tag is clicked, the final window.location.href is `devGuide/devGuide.html` instead of `devGuide/index.html` - so exact matching on the anchor tag will not work

--------------------------------
Also, I think I'm quite confused about the algorithm being described so far, so could I check if this is what you mean:
```js
const url = normalize(window.location.href)
if url ends wth `index.html`, do an exact match search
else:
    fallback to highlighting based on highlight-on attribute
    (where child or sibling means to highlight the 'highest-priority-by-number-of-groups-matched links')
```

Reply:
My original suggestion here https://github.com/MarkBind/markbind/pull/1379#discussion_r511595069:

The condition should be the opposite. The `url` variable should be the link, not `window.location.href` as well:

```js
const link = normalize(`<a href>` link)
if link ends wth `index.html`
  do an exact match search.
  If that fails,
    fallback to the relaxed search
    // because xx/index.html commonly is the (sub)home page of a whole bunch of other urls
else:
  only do exact search
```

---

The alternative I mentioned here https://github.com/MarkBind/markbind/pull/1379#discussion_r515591206:

```js
const link = normalize(`<a href>` link)
do an exact match search.
If that fails, **and** the mode of checking allows for falling back
  fallback to the relaxed search
  // because xx/index.html commonly is the (sub)home page of a whole bunch of other urls
```

This is also what's currently being done in your pr, if I read your code right.
The problem you encountered can be resolved by adding some sort of `highlight-on` attribute that allows controlling the mode of checking. (priority based checking is probably a bad idea as you've found, because we can have equal priority links)


Option 1 would likely result in less work here and on the author side, but:
- it is somewhat obscure to the author in that the highlighting behaviour depends on the link
- I'm not sure if there's any use cases that is impossible to accomodate to using this

Option 2 would need some more intervention from the author and implementation here, but should be the most flexible option

Reply:
> > `Hmm not sure I understood this, what does "not using the exact url" refer to? (wouldn't links as specified by <a href="..." /> always point to an exact url?)`
> 
> I was talking about the case where the href of an anchor tag doesn't match the _end result_ of `window.location.href` once it is clicked, for example:
> 
> 1. Suppose we have this entry in `layouts/devGuide/header.md`
> 
> ```
> <div tags="environment--dg">
> <li>
> <a href="{{baseUrl}}/devGuide/index.html" class="nav-link">DEVELOPER GUIDE</a>
> </li>
> </div>
> ```
> 
> 1. However, `devGuide/index.md` contains this code:
> 
> > <script>
> > window.location.href = "devGuide.html"
> > </script>
> 
> 1. So when the anchor tag is clicked, the final window.location.href is `devGuide/devGuide.html` instead of `devGuide/index.html` - so exact matching on the anchor tag will not work

So in this example, both options 1 & 2 would accomodate it without any intervention from the author
(the `<a href="{{baseUrl}}/devGuide/index.html" class="nav-link">DEVELOPER GUIDE</a>` is highlighted)

Reply:
Option 2 would be more future-proof; I believe it'll be better to go ahead with that